lib/levels/enemies/enemies-one.js

```diff
@@ -1,10 +1,13 @@
 const SimpleEnemy = require('../../simple-enemy');
-const QuickEnemy = require('../../quick-enemy');
 
-var enemiesOne = [
-  new SimpleEnemy({x: -100, y: 200}),
-  new SimpleEnemy({x: -200, y: 200}),
-  new SimpleEnemy({x: -300, y: 200})
-];
+let enemiesOne = {
+  createEnemies: function() {
+    return [
+      new SimpleEnemy({ x: -100, y: 200 }),
+      new SimpleEnemy({ x: -200, y: 200 }),
+      new SimpleEnemy({ x: -300, y: 200 })
+    ];
+  }
+};
```

Comment:
Object-oriented programming is weird. To quote friend-of-Turing, Ben
Orenstein, almost all advanced bugs and strange behavior are caused by the
unexpected mutation of state. It’s insane how many bugs get squashed in
Rails just by freezing objects and preventing mutation.

In your first approach, you were instantiating the enemies when declaring
the object. It’s likely that they are being changed and mutated as your
game goes on. When you have a function that returns, it is creating a fresh
batch that hasn’t been mutated.

On Sat, Apr 2, 2016 at 1:32 PM, Toni Rib <notifications@github.com> wrote:

> In lib/levels/enemies/enemies-one.js
> <https://github.com/ToniRib/game_time/pull/24#discussion_r58298434>:
>
> >
> > -var enemiesOne = [
> > -  new SimpleEnemy({x: -100, y: 200}),
> > -  new SimpleEnemy({x: -200, y: 200}),
> > -  new SimpleEnemy({x: -300, y: 200})
> > -];
> > +let enemiesOne = {
> > +  createEnemies: function() {
> > +    return [
> > +      new SimpleEnemy({ x: -100, y: 200 }),
> > +      new SimpleEnemy({ x: -200, y: 200 }),
> > +      new SimpleEnemy({ x: -300, y: 200 })
> > +    ];
> > +  }
> > +};
>
> Alright so my conclusion is this: JS is weird. If I wrap the enemy
> creation in a function and return the array, it works. So I'm going to
> update our other enemy files to match this. However, it would still be
> great to understand WHY this works over the other method.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/ToniRib/game_time/pull/24/files/b18b26bd9094860fecc38da7de99322513accb9c#r58298434>
>